### PR TITLE
Fuse operations and methods for mixed types

### PR DESCRIPTION
**Pull request by lbenet**
This PR began as an attempt to get rid of the use `fixorder`, which creates copies of Taylor1 objects. This has been accomplished to a certain extent using dot-operatotions and dot-functions. But I couldn't get rid of some cases. 

While working on this, #92 appeared followed by #94. I think I have solved this now by introducing new methods, and restricting more carefully some methods according to their methods.

### CODE REVIEW

**Comment by lbenet** [comment]
Pushed another commit to increase coverage

**Comment by lbenet** [comment]
To my surprise (at least in Julia 0.5), this PR seems to solve the cases described in #62 and #78 related to the order of loading `TaylorSeries` and another package. 

Here is an example:
```julia
julia> using TaylorSeries, ArbFloats

julia> t1=Taylor1([ArbFloat(2), ArbFloat(3)])
 2 + 3 t + 𝒪(t²)

julia> exp(t1)
 7.389056098930650227230427 + 22.16716829679195068169128 t + 𝒪(t²)
```

and the other way around

```julia
julia> using ArbFloats, TaylorSeries

julia> t1=Taylor1([ArbFloat(2), ArbFloat(3)])
 2 + 3 t + 𝒪(t²)

julia> exp(t1)
 7.389056098930650227230427 + 22.16716829679195068169128 t + 𝒪(t²)

```

@PerezHz @JeffreySarnoff @fhenneke Can you confirm this indeed the case?

**Comment by PerezHz** [comment]
Wow, quite unexpected! On my machine, the order of loading seems to be fixed for ArbFloats, ForwardDiff, and ValidatedNumerics!

**Comment by lbenet** [comment]
@dpsanders Do you want to take a look on this PR? (Sorry for making it so complicate.) Tests pass and have been enlarged to cover other situations (#94), including unexpected ones (#62 and #78).

**Comment by dpsanders** [comment]
Yes, but I won't be able to take a proper look at it until Sunday.
If you prefer to merge now then go ahead.

**Comment by lbenet** [comment]
Take your time. You have always good comments and insight! Thanks!

**Comment by dpsanders** [comment]
Hmm, I see that not using `fixorder` does complicate the code quite a lot.

It's also a bit unfortunate that what was now relatively simple code for `Taylor1` is mixed up with more complicated code for `TaylorN`?

**Comment by lbenet** [comment]
I agree, though your idea in #96 may help a little bit...

File: src/arithmetic.jl [hunk 1]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
-one{T<:Number}(a::TaylorN{T}) = TaylorN(one(T), a.order)
+zero(a::TaylorN) = TaylorN(zero(a.coeffs[1]), a.order)
+one(a::TaylorN) = TaylorN(one(a.coeffs[1]), a.order)
 
 
 
 ## Addition and substraction ##
 for f in (:+, :-)
-    for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
+    dotf = ifelse( f == :+, :(.+), :(.-))
```
Comment (dpsanders) [inline src/arithmetic.jl]: `dotf = Symbol(".", f)`

File: src/arithmetic.jl [hunk 2]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
         @eval begin
             ($f){T<:Number,S<:Number}(a::$T{T}, b::$T{S}) = $f(promote(a,b)...)
             function $f{T<:Number}(a::$T{T}, b::$T{T})
-                a, b = fixorder(a, b)
-                v = similar(a.coeffs)
-                @simd for i in eachindex(v)
-                    @inbounds v[i] = $f(a.coeffs[i], b.coeffs[i])
+                aord = a.order
+                bord = b.order
+                ac = view(a.coeffs, :)
```
Comment (dpsanders) [inline src/arithmetic.jl]: Why is this needed?

File: src/arithmetic.jl [hunk 3]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
+    dotf = ifelse( f == :+, :(.+), :(.-))
+    for T in (:Taylor1, :TaylorN)
         @eval begin
             ($f){T<:Number,S<:Number}(a::$T{T}, b::$T{S}) = $f(promote(a,b)...)
             function $f{T<:Number}(a::$T{T}, b::$T{T})
-                a, b = fixorder(a, b)
-                v = similar(a.coeffs)
-                @simd for i in eachindex(v)
-                    @inbounds v[i] = $f(a.coeffs[i], b.coeffs[i])
+                aord = a.order
```
Comment (dpsanders) [inline src/arithmetic.jl]: I would suggest using `la = length(a)` again.

File: src/arithmetic.jl [hunk 4]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
+                bord = b.order
+                ac = view(a.coeffs, :)
+                bc = view(b.coeffs, :)
+                if aord == bord
+                    v = similar(ac)
+                    v .= $dotf(ac, bc)
+                elseif aord < bord
+                    v = similar(bc)
+                    @inbounds v[1:aord+1] .= $dotf(ac[1:aord+1], bc[1:aord+1])
+                    @inbounds v[aord+2:bord+1] .= $f(bc[aord+2:bord+1])
```
Comment (dpsanders) [inline src/arithmetic.jl]: Use `end` instead of `bord+1`.

File: src/arithmetic.jl [hunk 5]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
-                    @simd for i in eachindex(coeffs)
-                        @inbounds coeffs[i] = $f(a.coeffs[i])
-                    end
-                    @inbounds coeffs[1] = $f(b, a.coeffs[1])
-                    return $T(coeffs, a.order)
-                end
+            ($f){T<:Number,S<:Number}(a::$T{T}, b::S) = $f(promote(a,b)...)
+            function $f{T<:Number}(a::$T{T}, b::T)
+                R = eltype(a.coeffs)
+                coeffs = Array{R}(a.order+1)
```
Comment (dpsanders) [inline src/arithmetic.jl]: Aren't these two lines just `coeffs = similar(a, a.order+1)`?

File: src/arithmetic.jl [hunk 6]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
+                coeffs .= a.coeffs
+                @inbounds coeffs[1] = $f(a.coeffs[1], b)
+                return $T(coeffs, a.order)
+            end
+            ($f){T<:Number,S<:Number}(b::S, a::$T{T}) = $f(promote(b,a)...)
+            function $f{T<:Number}(b::T, a::$T{T})
```
Comment (dpsanders) [inline src/arithmetic.jl]: Please add a new line before each new function definition.

File: src/arithmetic.jl [hunk 7]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
-                end
+            ($f){T<:Number,S<:Number}(a::$T{T}, b::S) = $f(promote(a,b)...)
+            function $f{T<:Number}(a::$T{T}, b::T)
+                R = eltype(a.coeffs)
+                coeffs = Array{R}(a.order+1)
+                coeffs .= a.coeffs
+                @inbounds coeffs[1] = $f(a.coeffs[1], b)
+                return $T(coeffs, a.order)
+            end
+            ($f){T<:Number,S<:Number}(b::S, a::$T{T}) = $f(promote(b,a)...)
```
Comment (dpsanders) [inline src/arithmetic.jl]: Shouldn't all these promotions be automatic for standard function like `+`?

File: src/arithmetic.jl [hunk 8]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
+        function $f{T<:NumberNotSeriesN}(a::HomogeneousPolynomial{T},
+                b::HomogeneousPolynomial{T})
+            @assert a.order == b.order
+            v = similar(a.coeffs)
+            v .= $dotf(a.coeffs, b.coeffs)
+            return HomogeneousPolynomial(v, a.order)
+        end
+        function $f(a::HomogeneousPolynomial)
+            v = similar(a.coeffs)
+            v .= $f(a.coeffs)
```
Comment (dpsanders) [inline src/arithmetic.jl]: These two lines can be reduced to `v = $f(a.coeffs)`.

File: src/arithmetic.jl [hunk 9]
```diff
@@ -73,88 +70,119 @@ function ones{T<:Number}(a::HomogeneousPolynomial{T}, order::Int)
...
                 a::TaylorN{Taylor1{T}}, b::Taylor1{S})
             @inbounds aux = $f(a.coeffs[1].coeffs[1], b)
-            R = typeof(aux)
-            coeffs = Array{HomogeneousPolynomial{Taylor1{R}}}(length(a.coeffs))
-            @simd for i in eachindex(coeffs)
-                @inbounds coeffs[i] = a.coeffs[i]
-            end
+            R = eltype(aux)
+            coeffs = Array{HomogeneousPolynomial{Taylor1{R}}}(a.order+1)
+            coeffs .= a.coeffs
```
Comment (dpsanders) [inline src/arithmetic.jl]: These two lines should be equivalent to a `convert`?

File: src/arithmetic.jl [hunk 10]
```diff
@@ -165,37 +193,39 @@ for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
...
+    end
+end
+
+for T in (:HomogeneousPolynomial, :TaylorN)
+    @eval begin
+        function *{T<:NumberNotSeries,S<:NumberNotSeries}(a::Taylor1{T}, b::$T{Taylor1{S}})
+            @inbounds aux = a * b.coeffs[1]
+            R = typeof(aux)
+            coeffs = Array{R}(length(b.coeffs))
+            coeffs .= a .* b.coeffs
```
Comment (dpsanders) [inline src/arithmetic.jl]: It feels like these three lines form a recurring pattern that should be factored out into a function somehow.

File: src/arithmetic.jl [hunk 11]
```diff
@@ -165,37 +193,39 @@ for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
...
+            return $T(coeffs, b.order)
         end
+        *{T<:NumberNotSeries,R<:NumberNotSeries}(b::$T{Taylor1{R}}, a::Taylor1{T}) = a * b
+        function *{T<:NumberNotSeries,S<:NumberNotSeries}(a::$T{T}, b::Taylor1{$T{S}})
+            @inbounds aux = a * b.coeffs[1]
+            R = typeof(aux)
+            coeffs = Array{R}(length(b.coeffs))
+            coeffs .= a .* b.coeffs
```
Comment (dpsanders) [inline src/arithmetic.jl]: There is a lot of repetition here with the previous definitions.

File: src/arithmetic.jl [hunk 12]
```diff
@@ -262,7 +292,7 @@ Inputs are the `kcoef`-th coefficient, and the vectors of the expansion coeffici
 """
 function mulHomogCoef{T<:Number}(kcoef::Int, ac::Array{T,1}, bc::Array{T,1})
     kcoef == 0 && return ac[1] * bc[1]
-    coefhomog = zero(T)
+    coefhomog = zero(ac[1])
```
Comment (dpsanders) [inline src/arithmetic.jl]: `zero(T)`.

File: src/auxiliary.jl [hunk 13]
```diff
@@ -8,27 +8,35 @@
 
 ## Auxiliary function ##
 
-function check_taylor1_order!{T<:Number}(coeffs::Array{T,1}, order::Int)
+"""
+    resize_coeffs1!{T<Number}(coeffs::Array{T,1}, order::Int)
```
Comment (lbenet) [inline src/auxiliary.jl]: I moved part of the constructor of `Taylor1` to this function to simplify the code. I'm not sure how to answer the question, since this function resizes the coefficients if you ask for a larger order. For instannce, you may call `Taylor1([0,1],10)` and at some point this function is called to resize the coefficients to size 11.

Comment (dpsanders) [inline src/auxiliary.jl]: My point is that there will no longer be any need to resize the coefficients, ever!

File: src/constructors.jl [hunk 14]
```diff
@@ -179,3 +182,11 @@ julia> TaylorN(Rational{Int},2)
 TaylorN{T<:Number}(::Type{T}, nv::Int; order::Int=get_order()) =
     return TaylorN( [HomogeneousPolynomial(T, nv)], order )
 TaylorN(nv::Int; order::Int=get_order()) = TaylorN(Float64, nv, order=order)
+
+
+# A `Number` which is not an `AbstractSeries`
+const NumberNotSeries = Union{setdiff(subtypes(Number), [AbstractSeries])...}
```
Comment (lbenet) [inline src/constructors.jl]: Somehow beyond my understanding, I think this is at the core why problems reported in #62 and #94 are solved.... or Julia-magic.

File: src/auxiliary.jl [hunk 15]
```diff
@@ -8,27 +8,35 @@
...
     lencoef = length(coeffs)
-    order = max(order, lencoef-1)
-    order == lencoef-1 && return nothing
+    order ≤ lencoef-1 && return nothing
     resize!(coeffs, order+1)
-    @simd for i = lencoef+1:order+1
-        @inbounds coeffs[i] = zero(coeffs[1])
-    end
-    nothing
+    coeffs[lencoef+1:order+1] .= zero(coeffs[1])
```
Comment (lbenet) [inline src/auxiliary.jl]: I guess it should work... I'll push another commit with this change.

File: src/auxiliary.jl (hunk 13 above)
Comment (lbenet) [inline src/auxiliary.jl]: I truly hope so!

File: src/evaluate.jl [hunk 16]
```diff
@@ -96,43 +96,42 @@ end
...
-
-    numVars = get_numvars()
-    @assert length(vals) == numVars
+# function evaluate{T<:Number}(a::HomogeneousPolynomial{T}, vals::Array{T,1} )
+function evaluate{T<:Number,S<:NumberNotSeriesN}(a::HomogeneousPolynomial{T},
+        vals::Array{S,1} )
+    @assert length(vals) == get_numvars()
     R = promote_type(T,S)
+    numVars = get_numvars()
     suma = convert(TaylorN{R}, a)
```
Comment (lbenet) [inline src/evaluate.jl]: Type stability; if you simply use assignment (`suma = a`) it may change type inside the loop.

File: src/evaluate.jl [hunk 17]
```diff
@@ -96,43 +96,42 @@ end
...
 
-    for nv = 1:numVars
+    @inbounds for nv = 1:numVars
         suma = horner(suma, (nv, vals[nv]))
     end
 
     return suma.coeffs[1].coeffs[1]
```
Comment (lbenet) [inline src/evaluate.jl]: Maybe. The way it is done here is that a `TaylorN` polynomial is evaluated substituting each variable, one by one; maybe it's not the best way of doing it...

File: src/evaluate.jl [hunk 18]
```diff
@@ -96,43 +96,42 @@ end
...
 end
 
-evaluate{T<:Number}(a::TaylorN{T}) = evaluate(a, zeros(T, get_numvars()))
+evaluate{T<:Number}(a::TaylorN{T}) = a.coeffs[1].coeffs[1]
```
Comment (lbenet) [inline src/evaluate.jl]: Evaluating a polynomial on zero is equivalent to the independent order zero) coefficient.

File: src/arithmetic.jl [hunk 19]
```diff
@@ -13,37 +13,36 @@ for T in (:Taylor1, :TaylorN)
     @eval begin
         =={T<:Number,S<:Number}(a::$T{T}, b::$T{S}) = ==(promote(a,b)...)
         function =={T<:Number}(a::$T{T}, b::$T{T})
-            a, b = fixorder(a, b)
-            @simd for i in eachindex(a.coeffs)
-                @inbounds (a.coeffs[i] == b.coeffs[i]) || return false
+            if a.order == b.order
+                return all( a.coeffs .== b.coeffs )
+            elseif a.order < b.order
+                res1 = all( a.coeffs[1:a.order+1] .== b.coeffs[1:a.order+1] )
```
Comment (lbenet) [inline src/arithmetic.jl]: For `Taylor1` this is essentially equivalent; for `TaylorN` this is a bit more involved. In any case, it's clearer the way you present it; I'll implement it.

File: src/arithmetic.jl [hunk 20]
```diff
@@ -13,37 +13,36 @@ for T in (:Taylor1, :TaylorN)
...
-    return true
+for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
+    @eval iszero(a::$T) = iszero(a.coeffs)
 end
 
-
 ## zero and one ##
-zero{T<:Number}(a::Taylor1{T}) = Taylor1(zero(a.coeffs[1]), a.order)
-one{T<:Number}(a::Taylor1{T}) = Taylor1(one(a.coeffs[1]), a.order)
+zero(a::Taylor1) = Taylor1(zero(a.coeffs[1]), a.order)
```
Comment (lbenet) [inline src/arithmetic.jl]: No. It's subtle, I agree, but things like `Taylor1{TaylorN{T}}` don't work correctly, because `zero(::Type{TaylorN{T}})` has no way of knowing about the order of the polynomial.

File: src/arithmetic.jl (hunk 1 above)
Comment (lbenet) [inline src/arithmetic.jl]: I'll implement it like that. Thanks!

File: src/arithmetic.jl (hunk 3 above)
Comment (lbenet) [inline src/arithmetic.jl]: I'll do it.

File: src/arithmetic.jl (hunk 2 above)
Comment (lbenet) [inline src/arithmetic.jl]: Not really needed, but the expressions were getting too long. Maybe with `la` things are easier to read.

File: src/arithmetic.jl (hunk 4 above)
Comment (lbenet) [inline src/arithmetic.jl]: I will. Thanks.

File: src/arithmetic.jl (hunk 5 above)
Comment (lbenet) [inline src/arithmetic.jl]: I think so, but there were subtleties related with the mixed types.

File: src/arithmetic.jl (hunk 7 above)
Comment (lbenet) [inline src/arithmetic.jl]: For `+` and `-` it's worth the code, since promotion would add lots of coefficients to zero.

File: src/arithmetic.jl (hunk 6 above)
Comment (lbenet) [inline src/arithmetic.jl]: I will! Sorry...

File: src/arithmetic.jl (hunk 8 above)
Comment (lbenet) [inline src/arithmetic.jl]: I guess so... Thanks for the catch!

File: src/arithmetic.jl (hunk 10 above)
Comment (lbenet) [inline src/arithmetic.jl]: I guess they appear like that to assure type stability.

File: src/arithmetic.jl (hunk 12 above)
Comment (lbenet) [inline src/arithmetic.jl]: Same as some comment above; the difference appears for `Taylor1{TaylorN{T}}`...

**Comment by lbenet** [comment]
I'll revise the code and push another commit.

**Comment by lbenet** [comment]
@dpsanders Thanks for your review! And, again, I'm sorry that this PR became so extended and messy.

I just pushed another commit which address many of your observations, but not all. As you suggest, this can be part of another commit. Once we get green lights, I think this should be merged.

Somewhere, which i can't find now, you also suggested that the tests should reflect more the file that they are actually testing. I will reorganize them as well, but in another PR.

**Comment by lbenet** [comment]
Got green lights. Merging.

**Comment by dpsanders** [comment]
Nice!
